web/pandas/pdeps/0007-copy-on-write.md

```diff
@@ -0,0 +1,575 @@
+
+Pandas, being built on NumPy, uses those concepts, and also exposes the behaviour
+consequences to its users. This basically means that pandas users, to understand the
+details of how indexing works, also need to understand those view / fancy indexing
+concepts of numpy.
+
+However, because DataFrames are not an array, the copy/view rules still differ from
+numpy's rules with current pandas. Slicing rows generally gives a view (following
+numpy), but slicing columns doesn't always give a view (this could be changed to match
+numpy however, see "Alternatives" 1b below). Fancy indexing rows (eg with a list of
+(positional) labels) gives a copy, but fancy indexing columns _could_ give a view
+(currently this gives a copy as well, but one of the "Alternatives" (1b) is to have this
+always return a view).
+
+The proposal in this document is to decouple the pandas user-facing behaviour from those
+numpy concepts. Creating a subset of a DataFrame with a slice or with a mask would
+behave in a similar way for the user (both return a new object and behave as a copy of
+the original). We still use the concept of views internally in pandas to optimize the
+implementation, but this becomes hidden from the user.
+
+## Alternatives
+
+The [original document](https://docs.google.com/document/d/1csGE4qigPR2vzmU2--jwURn3sK5k5eVewinxd8OUPk0/edit) and GitHub issue ([Proposal for future copy / view semantics in indexing operations - #36195](https://github.com/pandas-dev/pandas/issues/36195)) discussed several options for making the copy/view situation more consistent and clear:
+
+1. **Well-Defined copy/view rules:** ensure we have more consistent rules about which
+   operations result in a copy and which in a view, and then views result in mutating
+   the parent, copies not.
+   a. A minimal change would be to officialize the current behaviour. This comes down to
+      fixing some bugs and clearly documenting and testing which operations are views,
+      and which are copies.
+   b. An alternative would be to simplify the set of rules. For example: selecting
+      columns is always a view, subsetting rows is always a copy. Or: selecting columns
+      is always a view, subsetting rows as a slice is a view otherwise always a copy.
+
+2. **Copy-on-Write**: The setitem operation would check if it's a view on another
+   dataframe. If it is, then we would copy our data before mutating. (i.e. this
+   proposal)
+3. **Error-on-Write**: The setitem operation would check if it's a subset of another
+   dataframe (both view of copy). Only rather than copying in case of a view we would
+   raise an exception telling the user to either copy the data with
+   ``.copy_if_needed()`` (name TBD) or mark the frame as "a mutable view" with
+   ``.as_mutable_view()`` (name TBD).
+
+This document basically proposes an extended version of option 2 (Copy-on-Write). Some
+personal arguments in favor of Copy-on-Write compared to the other options:
```

Comment:
Feels odd to have the word "personal" in a PDEP (and a few lines down as well).

Reply:
Will rewrite theses sections